comm with js: code generation / `bs.deriving` (#28)
melange-re/documentation-site: docs/communicate-with-javascript.md

`identity` will make sure that compiler will do nothing about the returned
value. It is rarely used, but introduced here for debugging purposes.

> ## Deriving code from types

Author: Renamed to "Generate getters, setters and constructors" in https://github.com/melange-re/documentation-site/pull/28/commits/924286e9df03305dc924c82756a71b2e74907365

As we saw in a [previous section](#non-shared-data-types), there are some types

> in Melange that compile to values that are not easily manipulable from

Author: I don't think the suggested sentence is grammatically correct.

will generate something similar to the following code:

```ocaml
type action =
  | Click
  | Submit of string
  | Cancel

let click = (Click : action)
let submit param = (Submit param : action)
let cancel = (Cancel : action)
```

Which will result in the following JavaScript code after compilation:

```javascript
function submit(param_0) {
  return /* Submit */{
          _0: param_0
        };
}

var click = /* Click */0;

var cancel = /* Cancel */1;
```

Note the generated definitions are lower-cased, and they can be safely used from
JavaScript code. For example, if the above JavaScript generated code was located
in a `generators.js` file, the definitions can be used like this:

```javascript
const generators = require('./generators.js');

const hello = generators.submit("Hello");
const click = generators.click;
```

#### Conversion functions

Use `@bs.deriving jsConverter` on a variant type to create converter functions
that allow to transform back and forth between JavaScript integers and Melange
variant values.

There are a few differences with `@bs.deriving accessors`:

- `jsConverter` works with the `bs.as` attribute, while `accessors` does not

> - `jsConverter` does not support variant tags with payload, while `accessors`

Author: Do you want me to create an issue on melange-re/melange?

generated:

```ocaml
val actionToJs : action -> int

val actionFromJs : int -> action option
```

The `{ jsConverter = newType }` payload can also be passed to `@bs.deriving`
attribute when working with polymorphic variants.

### Records

#### Accessing fields

Use `@bs.deriving accessors` on a record type to create accessors for its record
field names.

```ocaml
type pet = { name : string } [@@bs.deriving accessors]

let pets = [| { name = "Brutus" }; { name = "Mochi" } |]

let () = pets |. Belt.Array.map name |. Js.Array2.joinWith "&" |. Js.log
```

Melange will generate a function for each field defined in the record. In this
case, a function `name` that allows to get that field from any record of type
`pet`:

```ocaml
let name (param : pet) = param.name
```

Considering all the above, the produced JavaScript will be:

```js
function name(param) {
  return param.name;
}

var pets = [
  {
    name: "Brutus"
  },
  {
    name: "Mochi"
  }
];

console.log(Belt_Array.map(pets, name).join("&"));
```

#### Convert records into abstract types

Usually, it’s recommended to use plain records when compiling to JavaScript

> objects (see the section on [binding to JavaScript

Author: tried something in https://github.com/melange-re/documentation-site/pull/28/commits/a1d961d6d653d182b49d2126d785a1fc0fce589b.

objects](#bind-to-javascript-objects)). But there's a specific case where

> records may not be enough. This happens when a record type has optional fields,
> and we want the resulting JavaScript objects to exclude those fields entirely.

Reviewer: ```suggestion
records may not be enough: Specifically, when we want to emit a JavaScript object
where some of the keys may be nullable, i.e. not present at all in the JavaScript
object. `
```

For instance, consider the following record:

```ocaml
type person = {
  name : string;
  age : int option;
}
```

> The specific use-case we are referring to arises when we have a Melange value
> such as `{ name = "John"; age = None }`, but we desire the resulting JavaScript
> object to be `{name: "Carl"}`.

Reviewer: ```suggestion
An example of this use-case would be expecting `{ name = "John"; age = None }`
to generate a JavaScript such as `{name: "Carl"}`, where the `age` key doesn't
appear.
```

> To address this case, the `@bs.deriving abstract` attribute can prove useful.
> When applied to a record type, this attribute removes the record type definition
> and replaces it with an abstract type. Additionally, it generates these
> functions:

Reviewer: ```suggestion
The `@bs.deriving abstract` attribute exists to solve this problem. When
present in a record type, `@bs.deriving abstract` makes the record definition
abstract and generates the following functions instead:
```

- A constructor function for creating values of the type
- Getters and setters for accessing the record fields

> Let’s see an example. Considering this Melange code:

Reviewer: ```suggestion
`@bs.deriving abstract` effectively models a record-shaped JavaScript object
exclusively through a set of (generated) functions derived from attribute annotations
on the OCaml type definition.

Let’s see an example. Considering this Melange code:
```

```ocaml
type person = {
  name : string;
  age : int; [@bs.optional]
}
[@@bs.deriving abstract]
```

> Melange will transform the `person` type into an abstract type and generate the
> constructor, getters, and setters functions. So we would get an interface like
> this:

Reviewer: ```suggestion
Melange will make the `person` type abstract and generate constructor, getter
and setter functions. In our example, the OCaml signature would look like this after
preprocessing:
```

```ocaml
type person

val person : name:string -> ?age:int -> unit -> person

val nameGet : person -> string

val ageGet : person -> int option
```

> The `person` constructor enables the creation of values for the `person` type.
> It is the only valid way to create such values because using literals directly
> like `{ name = "Alice"; age = None }` will no longer type check, due to the
> opaque nature of the abstract type definition.

Reviewer: ```suggestion
The `person` function can be used to create values of `person`. It is the only
possible way to create values of this type, since Melange makes it abstract.
Using literals like `{ name = "Alice"; age = None }` directly doesn't type check.
```

[private types](https://v2.ocaml.org/manual/privatetypes.html), which are
features inherited by Melange from OCaml.

When the record type has mutable fields, Melange will generate setter functions
for them. For example:

```ocaml
type person = {
  name : string;
  mutable age : int;
}
[@@bs.deriving abstract]

let alice = person ~name:"Alice" ~age:20 

let () = ageSet alice 21
```

This will generate:

```javascript
var alice = {
  name: "Alice",
  age: 20
};

alice.age = 21;
```

If the `mutable` keyword is omitted from the interface file, Melange will not
include the setter function in the module signature, preventing other modules
from mutating any values from the type.

Private types can be used to prevent Melange from creating the constructor
function. For example, if we define `person` type as private:

```ocaml
type person = private {
  name : string;
  age : int;
}
[@@bs.deriving abstract]
```

The accessors `nameGet` and `ageGet` will still be generated, but not the
constructor `person`. This is useful when binding to JavaScript objects while
preventing any Melange code from creating values of such type.

> ### Create JavaScript objects using external functions

Author: I am not sure where this section should be. It is not `bs.deriving` but it is kind of "code generation". One alternative would be to move it to "Bind to JS objects" section... 🤔

Reviewer: I'd probably move it to the "Bind to JS objects" section, indeed.

Author: https://github.com/melange-re/documentation-site/pull/28/commits/987495dc2496d926b22108832a2b38a342511b1f.

> We have already explored one approach to creating JavaScript object literals by

Reviewer: ```suggestion
We have already explored one approach for creating JavaScript object literals by
```

using [`Js.t` values and the `bs.obj` extension](#using-jst-objects).

> In addition to that approach, Melange offers the `bs.obj` attribute, which can

Reviewer: ```suggestion
Melange additionally offers the `bs.obj` attribute, which can
```

be used in combination with external functions to create JavaScript objects.
When these functions are called, they generate objects with fields corresponding
to the labeled arguments of the function.

If any of these labeled arguments are defined as optional and omitted during
function application, the resulting JavaScript object will exclude the

> corresponding fields. This feature allows for runtime object creation with
> dynamic selection of the subset of fields to include.

Reviewer: ```suggestion
corresponding fields. This allows to create runtime objects and control whether
optional keys are emitted at runtime.
```

For example, assuming we need to bind to a JavaScript object like this:

```js
var homeRoute = {
  type: "GET",
  path: "/",
  action: () => console.log("Home"),
  // options: ...
};
```

The first three fields are required and the `options` field is optional. You can
declare a binding function like:

```ocaml
external route :
  _type:string ->
  path:string ->
  action:(string list -> unit) ->
  ?options:< .. > ->
  unit ->
  _ = ""
  [@@bs.obj]
```

Note that the empty string at the end of the function is used to make it
syntactically valid. The value of this string is ignored by the compiler.

Since there is an optional argument `options`, an additional unlabeled argument

> of type `unit` is required. This allows the compiler to determine if the

Reviewer: The argument isn't really required, it's just a convenience that allows not having to apply every optional argument with `?foo:None`.

Feel free to leave it as is, though? it's a good recommendation, it's just not "required".

Author: Tried to rewrite in https://github.com/melange-re/documentation-site/pull/28/commits/04ef00465e6834d0d4bf6fbb990af8e25c51ef44.

optional argument has been omitted when the function is applied. More
information about optional arguments can be found in the [OCaml
manual](https://v2.ocaml.org/manual/lablexamples.html#s:optional-arguments).

The return type of the function should be left unspecified using the wildcard

> type `_`. Melange will automatically infer the type of the resulting JavaScript

Reviewer: I don't like this recommendation. At the very least it should be `< .. > Js.t`, which makes it a lot more explicit what's going on.

Author: It does not matter really. We can change the recommendation, but if you write 

```
external route :
  _type:string ->
  path:string ->
  action:(string list -> unit) ->
  ?options:< .. > ->
  unit ->
  int = ""
  [@@bs.obj]
```

`melc` will still compile just fine.

Author: So it seems a bit useless, as a user, to type more characters that do not really affect the outcome in any sense. Do you mean to recommend using `< .. > Js.t` as some sort of documentation?

object.

In the route function, the `_type` argument starts with an underscore. When
binding to JavaScript objects with fields that are reserved keywords in OCaml,
Melange allows the use of an underscore prefix for the labeled arguments. The
resulting JavaScript object will have the underscore removed from the field
names. This is only required for the `bs.obj` attribute, while for other cases,
the `bs.as` attribute can be used to rename fields.

If we call the function like this:

```ocaml

> let homeRoute = route ~__type:"GET" ~path:"/" ~action:(fun _ -> Js.log "Home") ()

Reviewer: There's an extra underscore in `~__type`, is it intentional? I didn't check how it renders.

Author: No, was a typo. Fixed in https://github.com/melange-re/documentation-site/pull/28/commits/584f30bb1af73a3921ab32dd4b28b0e443ec148a.

```

> We would get the generated JavaScript, which does not include the `options`
> field as required:

Reviewer: ```suggestion
We get the following JavaScript, which does not include the `options`
field since its argument wasn't present:
```
